File: src/cli/services/commands/explain.ts
// SPDX-License-Identifier: Apache-2.0

import { EmitMode } from "../../../loader/models";
import { listGenerationTargetSupport } from "../../../loader/service";
import {
  createTemplateExplanation,
  listModuleTemplates,
  listWizardTemplates,
} from "../../../wizard/service";
import {
  CLI_EXIT_CODE,
  CliExecutionInput,
  CliExecutionResult,
  CliGlobalFlags,
} from "../../models";
import { createEnvelope, createFailureResult } from "../envelope";

interface ExplainCommandData {
  schemaVersion: "1";
  selector: {
    kind:
      | "overview"
      | "project-template"
      | "module-template"
      | "generation-target";
    value: string;
  };
  summary: string;
  details: string[];
  available: {
    projectTemplates: string[];
    moduleTemplates: string[];
    generationTargets: EmitMode[];
  };
}

export function runExplainCommand(
  flags: CliGlobalFlags,
  _input: Omit<CliExecutionInput, "argv">,
): CliExecutionResult<ExplainCommandData> {
  try {
    const projectTemplates = listWizardTemplates();
    const moduleTemplates = listModuleTemplates();
    const generationTargets = listGenerationTargetSupport();
    const requestedTemplate = flags.template?.trim() || null;
    const requestedTarget = flags.target?.trim() || null;

    if (requestedTemplate && requestedTarget) {
      return createFailureResult(
        "explain",
        flags.format,
        CLI_EXIT_CODE.runtimeFailure,
        [
          "explain: use either --template or --target, but not both in the same command.",
        ],
        [
          {
            code: "explain.ambiguous-selector",
            severity: "error",
            message: "Use either --template or --target for explain, not both.",
          },
        ],
        {
          schemaVersion: "1",
          selector: { kind: "overview", value: "" },
          summary: "",
          details: [],
          available: {
            projectTemplates: projectTemplates.map((item) => item.id),
            moduleTemplates: moduleTemplates.map((item) => item.id),
            generationTargets: generationTargets.map((item) => item.emit),
          },
        },
      );
    }

    if (requestedTemplate) {
      const projectTemplate = projectTemplates.find(
        (item) => item.id === requestedTemplate,
      );
      if (projectTemplate) {
        const details = [
          projectTemplate.description,
          ...createTemplateExplanation(projectTemplate.id),
          `Default libraries: ${projectTemplate.defaultLibraries.join(", ") || "(none)"}`,
          `Default features: ${formatFeatureFlags(projectTemplate.defaultFeatures)}`,
        ];
        return createExplainSuccess(
          flags,
          {
            kind: "project-template",
            value: projectTemplate.id,
          },
          `${projectTemplate.label} (${projectTemplate.id})`,
          details,
          projectTemplates,
          moduleTemplates,
          generationTargets,
        );
      }

      const moduleTemplate = moduleTemplates.find(
        (item) => item.id === requestedTemplate,
      );
      if (moduleTemplate) {
        const details = [
          moduleTemplate.description,
          `Function prefix: ${moduleTemplate.functionPrefix}`,
          "Use this template with alp scaffold to generate a module source/header baseline.",
        ];
        return createExplainSuccess(
          flags,
          {
            kind: "module-template",
            value: moduleTemplate.id,
          },
          `${moduleTemplate.label} (${moduleTemplate.id})`,
          details,
          projectTemplates,
          moduleTemplates,
          generationTargets,
        );
      }

      return createFailureResult(
        "explain",
        flags.format,
        CLI_EXIT_CODE.runtimeFailure,
        [
          `explain: unknown template '${requestedTemplate}'. Run alp explain without selectors to list available topics.`,
        ],
        [
          {
            code: "explain.template-unknown",
            severity: "error",
            message: `Unknown template '${requestedTemplate}'.`,
          },
        ],
        {
          schemaVersion: "1",
          selector: { kind: "overview", value: requestedTemplate },
          summary: "",
          details: [],
          available: {
            projectTemplates: projectTemplates.map((item) => item.id),
            moduleTemplates: moduleTemplates.map((item) => item.id),
            generationTargets: generationTargets.map((item) => item.emit),
          },
        },
      );
    }

    if (requestedTarget) {
      const target = generationTargets.find(
        (item) => item.emit === requestedTarget,
      );
      if (!target) {
        return createFailureResult(
          "explain",
          flags.format,
          CLI_EXIT_CODE.runtimeFailure,
          [
            `explain: unknown generation target '${requestedTarget}'. Run alp explain without selectors to list available targets.`,
          ],
          [
            {
              code: "explain.target-unknown",
              severity: "error",
              message: `Unknown generation target '${requestedTarget}'.`,
            },
          ],
          {
            schemaVersion: "1",
            selector: { kind: "overview", value: requestedTarget },
            summary: "",
            details: [],
            available: {
              projectTemplates: projectTemplates.map((item) => item.id),
              moduleTemplates: moduleTemplates.map((item) => item.id),
              generationTargets: generationTargets.map((item) => item.emit),
            },
          },
        );
      }

      const details = [
        `Display name: ${target.displayName}`,
        `Output path: ${target.outputRelativePath}`,
        `Preview label: ${target.preview.label}`,
        `Preview language: ${target.preview.languageId}`,
      ];
      return createExplainSuccess(
        flags,
        {
          kind: "generation-target",
          value: target.emit,
        },
        `${target.displayName} (${target.emit})`,
        details,
        projectTemplates,
        moduleTemplates,
        generationTargets,
      );
    }

    const overviewDetails = [
      "Use --template to explain a project template (init) or module template (scaffold).",
      "Use --target to explain a generation output target.",
      `Project templates: ${projectTemplates.map((item) => item.id).join(", ")}`,
      `Module templates: ${moduleTemplates.map((item) => item.id).join(", ")}`,
      `Generation targets: ${generationTargets.map((item) => item.emit).join(", ")}`,
    ];

    return createExplainSuccess(
      flags,
      {
        kind: "overview",
        value: "all",
      },
      "ALP explain topics",
      overviewDetails,
      projectTemplates,
      moduleTemplates,
      generationTargets,
    );
  } catch (error) {
    const message =
      error instanceof Error
        ? error.message
        : "Unexpected CLI explain failure.";
    return createFailureResult(
      "explain",
      flags.format,
      CLI_EXIT_CODE.internalFailure,
      ["explain: internal failure", message],
      [
        {
          code: "explain.internal-failure",
          severity: "error",
          message,
        },
      ],
      {
        schemaVersion: "1",
        selector: { kind: "overview", value: "" },
        summary: "",
        details: [],
        available: {
          projectTemplates: [],
          moduleTemplates: [],
          generationTargets: [],
        },
      },
    );
  }
}

function createExplainSuccess(
  flags: CliGlobalFlags,
  selector: ExplainCommandData["selector"],
  summary: string,
  details: string[],
  projectTemplates: ReturnType<typeof listWizardTemplates>,
  moduleTemplates: ReturnType<typeof listModuleTemplates>,
  generationTargets: ReturnType<typeof listGenerationTargetSupport>,
): CliExecutionResult<ExplainCommandData> {
  const data: ExplainCommandData = {
    schemaVersion: "1",
    selector,
    summary,
    details,
    available: {
      projectTemplates: projectTemplates.map((item) => item.id),
      moduleTemplates: moduleTemplates.map((item) => item.id),
      generationTargets: generationTargets.map((item) => item.emit),
    },
  };

  const textLines: string[] = [];
  if (flags.format === "text") {
    textLines.push(`explain: ${summary}`);
    if (!flags.quiet) {
      for (const detail of details) {
        textLines.push(`- ${detail}`);
      }
    }
  }

  return {
    format: flags.format,
    exitCode: CLI_EXIT_CODE.success,
    textLines,
    envelope: createEnvelope(
      "explain",
      {
        root: null,
        boardYaml: null,
      },
      data,
      [],
      CLI_EXIT_CODE.success,
    ),
  };
}

function formatFeatureFlags(features: {
  wifi: boolean;
  mqtt: boolean;
  ble: boolean;
  tls: boolean;
}): string {
  return [
    `wifi=${features.wifi}`,
    `mqtt=${features.mqtt}`,
    `ble=${features.ble}`,
    `tls=${features.tls}`,
  ].join(" ");
}

File: cli-rs/crates/alp-cli/src/commands/explain.rs
// SPDX-License-Identifier: Apache-2.0
//! `alp explain` — describe project/module templates and generation targets.
//!
//! Mirrors TS `runExplainCommand`: `--template` explains an init/scaffold
//! template, `--target` explains a generation output target, and no selector
//! prints an overview. Supplying both is an error (exit 1), as is an unknown id.

use alp_core::wizard::{ModuleTemplateDefinition, WizardFeatureFlags, WizardTemplateDefinition};
use alp_core::wizard::{list_module_templates, list_wizard_templates};
use alp_core::{GenerationTargetSupport, list_generation_target_support};

use super::CommandRun;
use crate::cli::{ExplainArgs, GlobalArgs};
use crate::envelope::{Envelope, Issue, Project};
use crate::exit::ExitCode;

#[derive(serde::Serialize)]
struct Selector {
    kind: String,
    value: String,
}

#[derive(serde::Serialize)]
struct Available {
    #[serde(rename = "projectTemplates")]
    project_templates: Vec<String>,
    #[serde(rename = "moduleTemplates")]
    module_templates: Vec<String>,
    #[serde(rename = "generationTargets")]
    generation_targets: Vec<String>,
}

#[derive(serde::Serialize)]
struct ExplainData {
    #[serde(rename = "schemaVersion")]
    schema_version: String,
    selector: Selector,
    summary: String,
    details: Vec<String>,
    available: Available,
}

pub fn run(g: &GlobalArgs, args: &ExplainArgs) -> CommandRun {
    let project_templates = list_wizard_templates();
    let module_templates = list_module_templates();
    let generation_targets = list_generation_target_support();

    let available = || Available {
        project_templates: project_templates
            .iter()
            .map(|t| t.id.as_str().to_string())
            .collect(),
        module_templates: module_templates
            .iter()
            .map(|t| t.id.as_str().to_string())
            .collect(),
        generation_targets: generation_targets
            .iter()
            .map(|t| t.emit.to_string())
            .collect(),
    };

    let requested_template = args
        .template
        .as_deref()
        .map(str::trim)
        .filter(|s| !s.is_empty());
    let requested_target = g.target.as_deref().map(str::trim).filter(|s| !s.is_empty());

    if requested_template.is_some() && requested_target.is_some() {
        return failure(
            g,
            available(),
            Selector {
                kind: "overview".to_string(),
                value: String::new(),
            },
            "ambiguous-selector",
            "Use either --template or --target for explain, not both.",
            vec![
                "explain: use either --template or --target, but not both in the same command."
                    .to_string(),
            ],
        );
    }

    if let Some(requested) = requested_template {
        if let Some(pt) = project_templates
            .iter()
            .copied()
            .find(|t| t.id.as_str() == requested)
        {
            return success(
                g,
                available(),
                Selector {
                    kind: "project-template".to_string(),
                    value: pt.id.as_str().to_string(),
                },
                format!("{} ({})", pt.label, pt.id.as_str()),
                project_template_details(pt),
            );
        }
        if let Some(mt) = module_templates
            .iter()
            .copied()
            .find(|t| t.id.as_str() == requested)
        {
            return success(
                g,
                available(),
                Selector {
                    kind: "module-template".to_string(),
                    value: mt.id.as_str().to_string(),
                },
                format!("{} ({})", mt.label, mt.id.as_str()),
                module_template_details(mt),
            );
        }
        return failure(
            g,
            available(),
            Selector {
                kind: "overview".to_string(),
                value: requested.to_string(),
            },
            "template-unknown",
            &format!("Unknown template '{requested}'."),
            vec![format!(
                "explain: unknown template '{requested}'. Run alp explain without selectors to list available topics."
            )],
        );
    }

    if let Some(requested) = requested_target {
        let Some(target) = generation_targets.iter().find(|t| t.emit == requested) else {
            return failure(
                g,
                available(),
                Selector {
                    kind: "overview".to_string(),
                    value: requested.to_string(),
                },
                "target-unknown",
                &format!("Unknown generation target '{requested}'."),
                vec![format!(
                    "explain: unknown generation target '{requested}'. Run alp explain without selectors to list available targets."
                )],
            );
        };
        return success(
            g,
            available(),
            Selector {
                kind: "generation-target".to_string(),
                value: target.emit.to_string(),
            },
            format!("{} ({})", target.display_name, target.emit),
            generation_target_details(target),
        );
    }

    let overview = vec![
        "Use --template to explain a project template (init) or module template (scaffold)."
            .to_string(),
        "Use --target to explain a generation output target.".to_string(),
        format!(
            "Project templates: {}",
            project_templates
                .iter()
                .map(|t| t.id.as_str())
                .collect::<Vec<_>>()
                .join(", ")
        ),
        format!(
            "Module templates: {}",
            module_templates
                .iter()
                .map(|t| t.id.as_str())
                .collect::<Vec<_>>()
                .join(", ")
        ),
        format!(
            "Generation targets: {}",
            generation_targets
                .iter()
                .map(|t| t.emit)
                .collect::<Vec<_>>()
                .join(", ")
        ),
    ];
    success(
        g,
        available(),
        Selector {
            kind: "overview".to_string(),
            value: "all".to_string(),
        },
        "ALP explain topics".to_string(),
        overview,
    )
}

fn project_template_details(pt: &WizardTemplateDefinition) -> Vec<String> {
    let mut details = vec![pt.description.to_string()];
    details.extend(pt.explanation.iter().map(|s| (*s).to_string()));
    let libraries = if pt.libs.is_empty() {
        "(none)".to_string()
    } else {
        pt.libs.join(", ")
    };
    details.push(format!("Default libraries: {libraries}"));
    details.push(format!(
        "Default features: {}",
        format_feature_flags(pt.features.as_ref())
    ));
    details
}

fn module_template_details(mt: &ModuleTemplateDefinition) -> Vec<String> {
    vec![
        mt.description.to_string(),
        format!("Function prefix: {}", mt.function_prefix),
        "Use this template with alp scaffold to generate a module source/header baseline."
            .to_string(),
    ]
}

fn generation_target_details(target: &GenerationTargetSupport) -> Vec<String> {
    vec![
        format!("Display name: {}", target.display_name),
        format!("Output path: {}", target.output_relative_path),
        format!("Preview label: {}", target.preview_label),
        format!("Preview language: {}", target.preview_language_id),
    ]
}

fn format_feature_flags(features: Option<&WizardFeatureFlags>) -> String {
    let (wifi, mqtt, ble, tls) = match features {
        Some(f) => (f.wifi, f.mqtt, f.ble, f.tls),
        None => (false, false, false, false),
    };
    format!("wifi={wifi} mqtt={mqtt} ble={ble} tls={tls}")
}

fn null_project() -> Project {
    Project {
        root: None,
        board_yaml: None,
    }
}

fn success(
    g: &GlobalArgs,
    available: Available,
    selector: Selector,
    summary: String,
    details: Vec<String>,
) -> CommandRun {
    let text = if g.is_json() {
        Vec::new()
    } else {
        explain_text(&summary, &details, g)
    };
    let data = ExplainData {
        schema_version: "1".to_string(),
        selector,
        summary,
        details,
        available,
    };
    let json = g.is_json().then(|| {
        Envelope::new(
            "explain",
            null_project(),
            data,
            Vec::new(),
            ExitCode::Success.code(),
        )
        .to_json()
    });
    CommandRun {
        exit: ExitCode::Success,
        text,
        json,
    }
}

fn failure(
    g: &GlobalArgs,
    available: Available,
    selector: Selector,
    code: &str,
    issue_message: &str,
    text_lines: Vec<String>,
) -> CommandRun {
    let issues = vec![Issue {
        code: format!("explain.{code}"),
        severity: "error".to_string(),
        message: issue_message.to_string(),
    }];
    let data = ExplainData {
        schema_version: "1".to_string(),
        selector,
        summary: String::new(),
        details: Vec::new(),
        available,
    };
    let text = if g.is_json() { Vec::new() } else { text_lines };
    let json = g.is_json().then(|| {
        Envelope::new(
            "explain",
            null_project(),
            data,
            issues,
            ExitCode::RuntimeFailure.code(),
        )
        .to_json()
    });
    CommandRun {
        exit: ExitCode::RuntimeFailure,
        text,
        json,
    }
}

fn explain_text(summary: &str, details: &[String], g: &GlobalArgs) -> Vec<String> {
    let mut lines = vec![format!("explain: {summary}")];
    if !g.quiet {
        for detail in details {
            lines.push(format!("- {detail}"));
        }
    }
    lines
}
